crates/git_ui/src/branch_picker.rs

```diff
@@ -44,47 +54,114 @@ pub fn open(
+    }
 }
 
 impl BranchList {
-    pub fn new(
-        delegate: BranchListDelegate,
-        rem_width: f32,
-        window: &mut Window,
-        cx: &mut Context<Self>,
-    ) -> Self {
-        let picker = cx.new(|cx| Picker::uniform_list(delegate, window, cx));
-        let _subscription = cx.subscribe(&picker, |_, _, _, cx| cx.emit(DismissEvent));
+    fn new(project: Entity<Project>, style: BranchListStyle, rem_width: f32, cx: &mut App) -> Self {
+        let popover_handle = PopoverMenuHandle::default();
         Self {
-            picker,
+            project,
+            picker: None,
             rem_width,
-            _subscription,
+            popover_handle,
+            default_focus_handle: cx.focus_handle(),
+            style,
+            _subscription: None,
         }
     }
+
+    fn reload_branches(&mut self, window: &mut Window, cx: &mut Context<Self>) {
+        let project = self.project.clone();
+        let style = self.style;
+        cx.spawn_in(window, |this, mut cx| async move {
+            let delegate = BranchListDelegate::new(project, style, 20, &cx).await?;
+            let picker =
+                cx.new_window_entity(|window, cx| Picker::uniform_list(delegate, window, cx))?;
+
+            this.update(&mut cx, |branch_list, cx| {
+                let subscription =
+                    cx.subscribe(&picker, |_, _, _: &DismissEvent, cx| cx.emit(DismissEvent));
+
+                branch_list.picker = Some(picker);
+                branch_list._subscription = Some(subscription);
+
+                cx.notify();
+            })?;
+
+            anyhow::Ok(())
+        })
+        .detach_and_log_err(cx);
+    }
 }
 impl ModalView for BranchList {}
 impl EventEmitter<DismissEvent> for BranchList {}
 
 impl Focusable for BranchList {
     fn focus_handle(&self, cx: &App) -> FocusHandle {
-        self.picker.focus_handle(cx)
+        self.picker
+            .as_ref()
+            .map(|picker| picker.focus_handle(cx))
+            .unwrap_or_else(|| self.default_focus_handle.clone())
     }
 }
 
 impl Render for BranchList {
     fn render(&mut self, _: &mut Window, cx: &mut Context<Self>) -> impl IntoElement {
         v_flex()
             .w(rems(self.rem_width))
-            .child(self.picker.clone())
-            .on_mouse_down_out(cx.listener(|this, _, window, cx| {
-                this.picker.update(cx, |this, cx| {
-                    this.cancel(&Default::default(), window, cx);
+            .when_some(self.picker.clone(), |div, picker| {
+                div.child(picker.clone()).on_mouse_down_out({
+                    let picker = picker.clone();
+                    cx.listener(move |_, _, window, cx| {
+                        picker.update(cx, |this, cx| {
+                            this.cancel(&Default::default(), window, cx);
+                        })
+                    })
                 })
-            }))
+            })
+            .when_none(&self.picker, |div| {
```

Comment:
I don't think that we should have `.when_none` as an API.

In cases like this where we have two branches on the same value, it would be better to use a `.map` with a `match` / `if let Some(_)` inside.

Reply:
I think the symmetry with the pre-existing API is sufficient aesthetic justification for this addition.